src/scheduler/MpiWorld.cpp

```diff
@@ -1104,6 +1209,70 @@ void MpiWorld::initLocalQueues()
     }
 }
 
+std::shared_ptr<faabric::MPIMessage>
+MpiWorld::recvBatchReturnLast(int sendRank, int recvRank, int batchSize)
+{
+    std::shared_ptr<faabric::scheduler::UnackedMessageBuffer> umb =
+      getUnackedMessageBuffer(sendRank, recvRank);
+
+    // When calling from recv, we set the batch size to zero and work
+    // out the total here
+    auto firstNullMsg = std::find(umb->msgs.begin(), umb->msgs.end(), nullptr);
+    if (batchSize == 0) {
+        batchSize = std::distance(firstNullMsg, umb->msgs.end()) + 1;
+    }
+
+    // Work out whether the message is sent locally or from another host
+    assert(thisHost == getHostForRank(recvRank));
+    const std::string otherHost = getHostForRank(sendRank);
+    bool isLocal = otherHost == thisHost;
+
+    // Recv message: first we receive all messages for which there is an id
+    // in the unacknowleged buffer but no msg. Note that these messages
+    // (batchSize - 1) were `irecv`-ed before ours.
+    std::shared_ptr<faabric::MPIMessage> m;
+    auto it = firstNullMsg;
+    if (isLocal) {
+        // First receive messages that happened before us
+        for (int i = 0; i < batchSize - 1; i++) {
+            SPDLOG_TRACE("MPI - pending recv {} -> {}", sendRank, recvRank);
+            auto _m = getLocalQueue(sendRank, recvRank)->dequeue();
+
+            assert(_m != nullptr);
```

Comment:
This assertion should go inside `getLocalQueue`.

Reply:
But `getLocalQueue` does not interact with the message at all right? We need to call `dequeue()` on it. Granted that we only `enqueue()` items ourselves (in local `send()`, not in remote ones anymore), and we assert this before enqueueing, thus I thought an `assert` went well.